packages/@lwc/node-reactions/src/core/traverse.ts

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { assert, forEach, isUndefined } from '@lwc/shared';
+import { ReactionRecord, QualifyingReactionTypes } from '../types';
+import {
+    getConnectedRecordsForElement,
+    getDisconnectedRecordsForElement,
+    marker,
+} from './reactions';
+import { queueReactionRecord } from './reaction-queue';
+
+/**
+ * Queue qualifying reaction callbacks for a single node.
+ * If the node is a host element, this method will only queue callbacks for the host element.
+ * It does not process the host's shadow tree.
+ */
+function queueReactionsForSingleElement(
+    elm: Element,
+    reactionTypes: QualifyingReactionTypes,
+    reactionQueue: ReactionRecord[]
+) {
+    // Disconnected callback has to be processed before connected callback
+    if (reactionTypes & 2) {
+        const reactionRecords = getDisconnectedRecordsForElement(elm);
+        if (!isUndefined(reactionRecords)) {
+            queueReactionRecord(reactionQueue, reactionRecords);
+        }
+    }
+    if (reactionTypes & 1) {
+        const reactionRecords = getConnectedRecordsForElement(elm);
+        if (!isUndefined(reactionRecords)) {
+            queueReactionRecord(reactionQueue, reactionRecords);
+        }
+    }
+}
+
+const ShadowRootQuerySelectorAll = ShadowRoot.prototype.querySelectorAll;
+/**
+ * Process nodes in a shadow tree
+ */
+function queueReactionsForShadowRoot(
+    sr: ShadowRoot,
+    reactionTypes: QualifyingReactionTypes,
+    reactionQueue: ReactionRecord[]
+) {
+    queueReactionsForNodeList(
+        ShadowRootQuerySelectorAll.call(sr, `[${marker}]`),
+        reactionTypes,
+        reactionQueue
+    );
+}
+
+/**
+ * Process nodes in a NodeList of marked elements
+ */
+function queueReactionsForNodeList(
+    nodeList: NodeList,
+    reactionTypes: QualifyingReactionTypes,
+    reactionQueue: ReactionRecord[]
+) {
+    forEach.call(nodeList, node => {
+        queueReactionsForSingleElement(node, reactionTypes, reactionQueue);
+        // If node has a shadow tree, process its shadow tree
+        queueReactionsForShadowRoot((node as Element).shadowRoot!, reactionTypes, reactionQueue);
```

Comment:
I don't see how you can ensure that the `node` has an associated `ShadowRoot`.

Reply:
So the NodeList is always the result of querying for the marked nodes, we can be sure it always results a NodeListOf<Element>. Will update the type definition for the arguments.